src/EventStore.Core/Services/SubscriptionMetric.cs

```diff
@@ -0,0 +1,74 @@
+#nullable enable
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.Metrics;
+using EventStore.Core.Messages;
+
+namespace EventStore.Core.Services;
+
+public class SubscriptionMetric {
+	private readonly ConcurrentDictionary<Guid, MonitoringMessage.SubscriptionStats> _subscriptionStats;
+
+	public SubscriptionMetric(Meter meter, string name) {
+		_subscriptionStats = new();
+
+		meter.CreateObservableUpDownCounter($"{name}-completed", Completed);
+		meter.CreateObservableUpDownCounter($"{name}-subscription-position", SubscriptionPosition);
+		meter.CreateObservableUpDownCounter($"{name}-stream-position", StreamPosition);
+	}
+
+	public void Add(Guid subscriptionId, string? streamName, long endOfStream) =>
+		_subscriptionStats.TryAdd(subscriptionId, new(subscriptionId, streamName, EndOfStream: endOfStream));
+
+	public void Remove(Guid subscriptionId) => _subscriptionStats.TryRemove(subscriptionId, out _);
+
+	public void RecordEvent(string? streamName, long commitPosition, long eventPosition) {
+		foreach (var (_, stats) in _subscriptionStats) {
+			if (stats.StreamName == streamName) {
+				_subscriptionStats.TryUpdate(stats.SubscriptionId, stats with {
+					EndOfStream = eventPosition
+				}, stats);
+			} else if (stats.StreamName is null) {
+				_subscriptionStats.TryUpdate(stats.SubscriptionId, stats with {
+					EndOfStream = commitPosition
+				}, stats);
+			}
+		}
+	}
+
+	public void ProcessEvent(string? streamName, long commitPosition, long eventPosition) {
+		foreach (var (_, stats) in _subscriptionStats) {
+			if (stats.StreamName == streamName) {
+				_subscriptionStats.TryUpdate(stats.SubscriptionId, stats with {
+					SubscriptionPosition = eventPosition
+				}, stats);
+			} else if (stats.StreamName is null) {
+				_subscriptionStats.TryUpdate(stats.SubscriptionId, stats with {
+					SubscriptionPosition = commitPosition
+				}, stats);
+			}
+		}
+	}
```

Comment:
- similarly, quite expensive to do for every event that we send to a subscription?
- also we probably want to update the specific subscription that processed this event or all the subscriptions will have the same position